presto-main/src/test/java/com/facebook/presto/operator/TestMergeOperator.java

```diff
@@ -0,0 +1,404 @@
+                .row(26, 2, 5)
+                .row(27, 1, 4)
+                .row(41, 2, 5)
+                .row(55, 1, 2)
+                .row(56, 1, 2)
+                .row(66, 1, 3)
+                .row(77, 1, 4)
+                .row(88, 1, 3)
+                .row(88, 1, 3)
+                .row(89, 1, 3)
+                .row(89, 1, 3)
+                .row(90, 1, 3)
+                .row(91, 1, 4)
+                .row(92, 2, 5)
+                .row(93, 1, 2)
+                .row(94, 1, 3)
+                .row(95, 1, 4)
+                .row(97, 1, 3)
+                .row(98, 2, 2)
+                .row(99, 2, 2)
+                .row(99, 1, 1)
+                .row(100, 1, 7)
+                .row(100, 1, 6)
+                .row(101, 1, 4)
+                .row(202, 1, 3)
+                .row(399, 2, 2)
+                .row(400, 1, 1)
+                .row(401, 1, 7)
+                .row(402, 1, 6)
+                .build();
+        outputPages.addAll(pullAvailablePages(operator));
+
+        MaterializedResult actual = toMaterializedResult(operator.getOperatorContext().getSession(), types, outputPages);
+        MaterializedResult expected = toMaterializedResult(operator.getOperatorContext().getSession(), types, expectedPages);
+        assertEquals(actual, expected);
+
+        assertOperatorIsUnblocked(operator);
+        assertTrue(operator.isFinished());
+
+        assertNull(operator.getOutput());
+
+        operator.close();
+    }
+
+    private MergeOperator createMergeOperator(List<Type> sourceTypes, List<Integer> outputChannels, List<Integer> sortChannels, List<SortOrder> sortOrder)
+    {
+        MergeOperator.MergeOperatorFactory factory = new MergeOperator.MergeOperatorFactory(
+                operatorId.getAndIncrement(),
+                new PlanNodeId("plan_node_id" + operatorId.getAndIncrement()),
+                exchangeClientSupplier,
+                serdeFactory,
+                new SimpleMergeSortComparator.Factory(),
+                sourceTypes,
+                outputChannels,
+                sortChannels,
+                sortOrder);
+        DriverContext driverContext = createTaskContext(executor, TEST_SESSION)
+                .addPipelineContext(0, true, true)
+                .addDriverContext();
+        return (MergeOperator) factory.createOperator(driverContext);
+    }
+
+    private static Split createRemoteSplit(String taskId)
+    {
+        return new Split(ExchangeOperator.REMOTE_CONNECTOR_ID, new RemoteTransactionHandle(), new RemoteSplit(URI.create("http://localhost/" + taskId)));
+    }
+
+    private static List<Page> pullAvailablePages(Operator operator)
+            throws InterruptedException
+    {
+        // read expected pages or until 10 seconds has passed
+        long endTime = System.nanoTime() + TimeUnit.SECONDS.toNanos(10);
+        List<Page> outputPages = new ArrayList<>();
+
+        boolean unblocked = false;
+        while (System.nanoTime() < endTime) {
```

Comment:
Why don't you extract this bit to the outer loop and put both loops inside (`System.nanoTime() < endTime` is a condition in both `while` loops).
You can leave the first loop only when timeout is reached or `operator.isFinished()` or `operator.isBlocked().isDone()` but right after the first loop you check if timeout is not reached AND whether `operator.isBlocked().isDone()`. Why not merge those two loops to avoid these misleading conditions?